File: day:71/findMedianSortedArrays.js
function merge_sorted_arrays(nums1, nums2) {
    /* --- Brute force method ------------------------- */
    let result = [];
    let [i, j, k] = [0 , 0, 0];

    while(i < nums1.length && j < nums2.length) {
        if(nums1[i] <= nums2[j]) {
            result[k] = nums1[i];
            i += 1;
            k += 1;
        }
        else {
            result[k] = nums2[j];
            j += 1;
            k += 1;
        }
    }
    if(i < nums1.length) {
        for(let j = i; j < nums1.length; j++) {
            result[k] = nums1[j];
            k += 1;
        }
    }
    if(j < nums2.length) {
        for(let i = j; i < nums2.length; i++) {
            result[k] = nums2[i];
            k += 1;
        }
    }

    return result;
}
function median(num) {
    // odd length
    if(num.length % 2 === 1) {
        return num[Math.floor(num.length/2)];
    }
    else {
        // even length
        let mid = Math.floor(num.length/2);
        return (num[mid - 1] + num[mid]) /2 ;
    }
}
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number}
 */
var findMedianSortedArrays_bruteforce = function(nums1, nums2) {
    let result = merge_sorted_arrays(nums1, nums2);
    return median(result);
};
const findMedianSortedArrays = (nums1, nums2) => {
    // 1. Find the shorter array of the two lists and make it nums1
    // 2. Find a pivot point in nums1 (using binary search)
    // 3. pX + pY = (x + y + 1)/2, where x = len(nums1) and y = len(nums2)
    if(nums1.length > nums2.length) {
        return findMedianSortedArrays(nums2, nums1);
    }
    let [x, y] = [nums1.length, nums2.length];
    let [low, high] = [0, x];

    while (low <= high) {
        let pX = Math.floor((low + high)/2);
        let pY = Math.floor((x + y + 1)/2) - pX;

        let maxLeftX = (pX === 0)? -Infinity: nums1[pX - 1];
        let minRightX = (pX === x)? Infinity: nums1[pX];
        let maxLeftY = (pY === 0)? -Infinity: nums2[pY - 1];
        let minRightY = (pY === y)? Infinity: nums2[pY];
        
        // Found the required condition
        if(maxLeftX <= minRightY && maxLeftY <= minRightX ) {
            let maxLeft = Math.max(maxLeftX, maxLeftY);
            let minRight = Math.min(minRightX, minRightY);
           // for even length
            if((x + y) % 2 === 0) {
                return (maxLeft + minRight)/2.0;
            }
           // for odd length
            else {
                return maxLeft;
            }
        }
        else if(maxLeftX > minRightY) {
            high = pX - 1;
        }
        else {
            low = pX + 1
        }
    }
}

File: day:71/findMedianSortedArrays.rs
impl Solution {
    pub fn find_median_sorted_arrays(a: Vec<i32>, b: Vec<i32>) -> f64 {
        let nums1: Vec<i32>;
    let nums2: Vec<i32>;
    if a.len() < b.len() {
        nums1 = a;
        nums2 = b;
    }
    else {
        nums1 = b;
        nums2 = a;
    }
    /*
        x = len(nums1), y = len(nums2)
        Px + Py = (x + y)/2
        maxLeft_x = -Infinity if Px == 0 else nums1[Px - 1]
        minRight_x = Infinity if Px == x else nums1[Px]
        maxRight_x = -Infinity if Py == 0 else nums2[Py - 1]
        minRight_x = Infinity if Py == y else nums2[Py]

        if even,
            median = max(maxLeft_x, maxLeft_y) + min(minLeft_x, minRight_x) / 2;
        else,
            median = max(maxLeft_x, maxLeft_y);
     */

    let (x, y) = (nums1.len() as i32, nums2.len() as i32);
    let (mut low, mut high) = (0 as i32, x);

    while low <= high {
        let mut px = ((low + high) as f64 / 2.0).floor() as usize;
        let mut py = (((x + y + 1) as f64 / 2.0).floor() as i32 - px as i32) as usize;

        let max_left_x = if px == 0 { std::i32::MIN} else {nums1[px - 1]};
        let min_right_x = if px == x as usize {std::i32::MAX} else {nums1[px]};
        let max_left_y = if py == 0 {std::i32::MIN} else {nums2[py - 1]};
        let min_right_y = if py == y as usize {std::i32::MAX} else {nums2[py]};

        if max_left_x <= min_right_y && max_left_y <= min_right_x {
            let max_left = if max_left_x > max_left_y {max_left_x} else {max_left_y};
            let min_left = if min_right_x < min_right_y {min_right_x} else {min_right_y};

            return if (x + y) % 2 == 0 {
                ((max_left + min_left) as f64 / 2.0) as f64
            } else {
                max_left as f64
            }
        }
        else if max_left_x > min_right_y {
            high = (px - 1) as i32;
        }
        else {
            low = (px + 1) as i32;
        }
    }
    return 0.0;
    }
}
